### PR TITLE
feature: support docker proxy

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: Yue Zhang <huaihuan.zy@alibaba-inc.com>

### Ⅰ. Describe what this PR does
to #64

### CODE REVIEW

File: pkg/runtime-manager/server/docker/server.go [hunk 1]
```diff
@@ -0,0 +1,314 @@
+package docker
```
Comment (jasonliu747) [inline pkg/runtime-manager/server/docker/server.go:1]: image

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

**Review by jasonliu747 — CHANGES_REQUESTED** [review]
please resolve conflicts.

File: pkg/runtime-manager/server/docker/server.go [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+	wr.Header().Set("Content-Length", strconv.Itoa(len(string(respBytes))))
+	wr.WriteHeader(resp.StatusCode)
+
+	written, err := io.Copy(wr, bytes.NewReader(respBytes))
+	if err != nil {
+		klog.Errorf("io copy error, written: %d, err: %v", written, err)
+	}
+}
+
+func (d *RuntimeManagerDockerServer) ServeHTTP(wr http.ResponseWriter, req *http.Request) {
```
Comment (jasonliu747) [inline pkg/runtime-manager/server/docker/server.go:86]: this function is waayyyyyy too long, and has too many if-else statements.

**Review by jasonliu747 — CHANGES_REQUESTED** [review]

File: pkg/runtime-manager/server/docker/server.go [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+		if err != nil {
+			http.Error(wr, "Failed to ready req body ", http.StatusInternalServerError)
+			return
+		}
+		config := &types2.ConfigWrapper{}
+		if err := json.Unmarshal(reqBytes, config); err != nil {
+			http.Error(wr, "Failed to Unmarshal req body to docker config ", http.StatusInternalServerError)
+			return
+		}
+		config.ContainerName = req.URL.Query()["name"][0]
```
Comment (hormes) [inline pkg/runtime-manager/server/docker/server.go:104]: Query()["name"][0] Check for correctness?

File: pkg/runtime-manager/server/docker/server.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+	written, err := io.Copy(wr, bytes.NewReader(respBytes))
+	if err != nil {
+		klog.Errorf("io copy error, written: %d, err: %v", written, err)
+	}
+}
+
+func (d *RuntimeManagerDockerServer) ServeHTTP(wr http.ResponseWriter, req *http.Request) {
+	if !strings.Contains(req.URL.Path, "containers") {
+		d.Direct(wr, req)
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/runtime-manager/server/docker/server.go]: fix it

File: pkg/runtime-manager/server/docker/server.go [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+	}
+
+	ctx := context.TODO()
+	reqCopy := req.Clone(ctx)
+	if strings.Contains(req.URL.Path, "create") {
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/runtime-manager/server/docker/server.go]: fix it

File: pkg/runtime-manager/server/docker/server.go (hunk 2 above)
Comment (ZYecho) [reply pkg/runtime-manager/server/docker/server.go:56]: fix it

File: pkg/runtime-manager/resource-executor/container.go [hunk 6]
```diff
@@ -90,6 +92,34 @@ func (c *ContainerResourceExecutor) ParseRequest(req interface{}) error {
...
+					Name: request.ContainerName,
+				},
+				ContainerAnnotations: request.Labels,
+				ContainerResources:   HostConfigToResource(request.HostConfig),
+			},
+		}
+	case *string:
+		return c.loadContainerInfoFromStore(*request, "Docker")
+	case *container.UpdateConfig:
+		//c.loadContainerInfoFromStore()
```
Comment (jasonliu747) [inline pkg/runtime-manager/resource-executor/container.go:115]: remove these codes if not needed anymore

File: pkg/runtime-manager/resource-executor/pod.go [hunk 7]
```diff
@@ -100,24 +113,38 @@ func (p *PodResourceExecutor) ParsePod(podsandbox *runtimeapi.PodSandbox) error
...
 
 func (p *PodResourceExecutor) ResourceCheckPoint(response interface{}) error {
-	runPodSandboxResponse, ok := response.(*runtimeapi.RunPodSandboxResponse)
-	if !ok || p.GetRunPodSandboxHookRequest() == nil {
-		return fmt.Errorf("no need to checkpoint resource %v %v", response, p.GetRunPodSandboxHookRequest())
-	}
-	err := store.WritePodSandboxInfo(runPodSandboxResponse.PodSandboxId, &p.PodSandboxInfo)
-	if err != nil {
-		return err
+	//runPodSandboxResponse, ok := response.(*runtimeapi.RunPodSandboxResponse)
```
Comment (jasonliu747) [inline pkg/runtime-manager/resource-executor/pod.go:116]: remove these codes if not needed anymore

**Review by jasonliu747 — CHANGES_REQUESTED** [review]
Please fix thoses CI errors and run `go mod vendor` again.

File: pkg/runtime-manager/server/docker/handler.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	err = resourceExecutor.ParseRequest(config)
+	if err != nil {
+		klog.Errorf("Failed to parse req to local store, err: %v", err)
+		http.Error(wr, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	// 3. send req to pre-hook
+	runtimeHookPath := types.NoneRuntimeHookPath
+	if response, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, types.PreHook, resourceExecutor.GenerateHookRequest()); err != nil {
+		klog.Errorf("fail to call hook server %v", err)
```
Comment (eahydra) [inline pkg/runtime-manager/server/docker/handler.go:64]: missing return error response?

Comment (eahydra) [reply pkg/runtime-manager/server/docker/handler.go:64]: please add UT for these files.

File: pkg/runtime-manager/server/docker/handler.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+		return
+	}
+
+	runtimeHookPath := types.NoneRuntimeHookPath
+	if runtimeResourceType == resource_executor.RuntimeContainerResource {
+		runtimeHookPath = types.UpdateContainerResources
+	}
+
+	if response, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, types.PreHook, resourceExecutor.GenerateHookRequest()); err != nil {
+		klog.Errorf("fail to call hook server %v", err)
```
Comment (eahydra) [inline pkg/runtime-manager/server/docker/handler.go:199]: missing return error?

File: pkg/runtime-manager/server/docker/server.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+func (d *RuntimeManagerDockerServer) Name() string {
+	return "RuntimeManagerDockerServer"
+}
+
+func NewRuntimeManagerDockerServer() *RuntimeManagerDockerServer {
+	criInterceptor := &RuntimeManagerDockerServer{
+		dispatcher: dispatcher.NewRuntimeDispatcher(),
+	}
+	criInterceptor.router = map[*regexp.Regexp]func(context.Context, http.ResponseWriter, *http.Request){
+		regexp.MustCompile("^/(v\\d\\.\\d+/)?containers(/\\w+)?/update$"): criInterceptor.HandleUpdateContainer,
```
Comment (eahydra) [inline pkg/runtime-manager/server/docker/server.go:55]: Can we replace this custom regular implementation with a more mature route package?

**Review by eahydra — CHANGES_REQUESTED** [review]

File: pkg/runtime-manager/resource-executor/container.go [hunk 11]
```diff
@@ -14,16 +14,18 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package cri
+package resource_executor
```
[the remark this answers is not in the document]
Comment (ZYecho) [reply pkg/runtime-manager/resource-executor/container.go]: in next pr

**Review by jasonliu747 — CHANGES_REQUESTED** [review]
image

File: pkg/runtime-manager/server/docker/server.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		klog.Errorf("failed to get docker server info, err: %v", err)
+		return err
+	}
+	d.cgroupDriver = info.CgroupDriver
+	return nil
+}
+
+func (d *RuntimeManagerDockerServer) Run() error {
+
```
Comment (jasonliu747) [inline pkg/runtime-manager/server/docker/server.go:167]: Redundant line

File: cmd/runtime-manager/main.go [hunk 13]
```diff
@@ -40,16 +43,22 @@ func main() {
 	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
 	pflag.Parse()
 
+	dir, _ := filepath.Split(options.RuntimeManagerEndpoint)
+	_ = os.Mkdir(dir, 0777)
```
Comment (jasonliu747) [inline cmd/runtime-manager/main.go:47]: is it guaranteed to succeed?

Comment (stormgbs) [reply cmd/runtime-manager/main.go:47]: 1. Handle error;
2. `os.MkdirAll` may be better;

File: pkg/runtime-manager/server/docker/handler.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+	runtimeHookPath := config.NoneRuntimeHookPath
+	var hookReq interface{}
+	if containerMeta != nil {
+		runtimeHookPath = config.StartContainer
+		hookReq = containerMeta.GetContainerResourceHookRequest()
+	}
+
+	// no need to care about the resp
+	if _, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, hookReq); err != nil {
+		klog.Errorf("fail to call pre start container hook server %v", err)
```
Comment (stormgbs) [inline pkg/runtime-manager/server/docker/handler.go:177]: fail -> Failed

File: pkg/runtime-manager/server/docker/handler.go [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+		http.Error(wr, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	d.Direct(wr, req)
+
+	if _, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, config.PostHook, hookReq); err != nil {
+		klog.Errorf("fail to call post start container hook server %v", err)
```
Comment (stormgbs) [inline pkg/runtime-manager/server/docker/handler.go:185]: fail -> Failed

File: pkg/runtime-manager/server/docker/handler.go [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+	containerMeta := store.GetContainerInfo(containerID)
+	runtimeHookPath := config.NoneRuntimeHookPath
+	if containerMeta != nil {
+		runtimeHookPath = config.UpdateContainerResources
+		hookReq = containerMeta.GetContainerResourceHookRequest()
+	}
+
+	response, err := d.dispatcher.Dispatch(ctx, runtimeHookPath, config.PreHook, hookReq)
+	if err != nil {
+		klog.Errorf("fail to call pre update hook server %v", err)
```
Comment (stormgbs) [inline pkg/runtime-manager/server/docker/handler.go:253]: Failed

File: pkg/runtime-manager/server/docker/server.go [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+	err := d.failOver()
+	if err != nil {
+		//FIXME: need to panic?
+		klog.Errorf("Failed to backup container info from backend, err: %v", err)
+	}
+
+	lis, err := net.Listen("unix", options.RuntimeManagerEndpoint)
+	if err != nil {
+		klog.Fatal("fail to create the lis %v", err)
```
Comment (stormgbs) [inline pkg/runtime-manager/server/docker/server.go:191]: Failed

File: pkg/runtime-manager/server/docker/handler.go [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package docker
+
+import (
```
Comment (hormes) [inline pkg/runtime-manager/server/docker/handler.go:19]: please create an issue to add tests for server and handler

Comment (ZYecho) [reply pkg/runtime-manager/server/docker/handler.go:19]: https://github.com/koordinator-sh/koordinator/issues/193

**Pull request merged**